File: code/redis/examples/session_manager.go
package main

/*
Redis セッション管理システム

使用方法:
    manager := NewRedisSessionManager("localhost:6379", 3600)
    sessionID, _ := manager.CreateSession(1001, map[string]interface{}{"theme": "dark"})
    sessionData, _ := manager.GetSession(sessionID)
*/

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/go-redis/redis/v8"
	"github.com/google/uuid"
)

var ctx = context.Background()

type RedisSessionManager struct {
	client     *redis.Client
	defaultTTL time.Duration
}

func NewRedisSessionManager(addr string, ttlSeconds int) *RedisSessionManager {
	client := redis.NewClient(&redis.Options{
		Addr: addr,
	})

	return &RedisSessionManager{
		client:     client,
		defaultTTL: time.Duration(ttlSeconds) * time.Second,
	}
}

func (m *RedisSessionManager) CreateSession(userID int, userData map[string]interface{}) (string, error) {
	sessionID := uuid.New().String()
	sessionKey := fmt.Sprintf("session:%s", sessionID)

	// セッションデータ構築
	now := time.Now().Format(time.RFC3339)
	userDataJSON, err := json.Marshal(userData)
	if err != nil {
		return "", err
	}

	sessionData := map[string]interface{}{
		"user_id":       userID,
		"created_at":    now,
		"last_accessed": now,
		"data":          string(userDataJSON),
	}

	// セッションデータ保存（Hash使用）
	if err := m.client.HSet(ctx, sessionKey, sessionData).Err(); err != nil {
		return "", err
	}

	// TTL設定
	if err := m.client.Expire(ctx, sessionKey, m.defaultTTL).Err(); err != nil {
		return "", err
	}

	// ユーザーのアクティブセッション管理（Set使用）
	userSessionsKey := fmt.Sprintf("user_sessions:%d", userID)
	if err := m.client.SAdd(ctx, userSessionsKey, sessionID).Err(); err != nil {
		return "", err
	}
	if err := m.client.Expire(ctx, userSessionsKey, m.defaultTTL).Err(); err != nil {
		return "", err
	}

	return sessionID, nil
}

func (m *RedisSessionManager) GetSession(sessionID string) (map[string]string, error) {
	sessionKey := fmt.Sprintf("session:%s", sessionID)
	sessionData, err := m.client.HGetAll(ctx, sessionKey).Result()
	if err != nil {
		return nil, err
	}

	if len(sessionData) == 0 {
		return nil, nil
	}

	// 最終アクセス時刻更新
	now := time.Now().Format(time.RFC3339)
	if err := m.client.HSet(ctx, sessionKey, "last_accessed", now).Err(); err != nil {
		return nil, err
	}
	if err := m.client.Expire(ctx, sessionKey, m.defaultTTL).Err(); err != nil {
		return nil, err
	}

	return sessionData, nil
}

func (m *RedisSessionManager) UpdateSession(sessionID string, data map[string]interface{}) error {
	sessionKey := fmt.Sprintf("session:%s", sessionID)
	dataJSON, err := json.Marshal(data)
	if err != nil {
		return err
	}

	if err := m.client.HSet(ctx, sessionKey, "data", string(dataJSON)).Err(); err != nil {
		return err
	}
	return m.client.Expire(ctx, sessionKey, m.defaultTTL).Err()
}

func (m *RedisSessionManager) DestroySession(sessionID string) error {
	sessionKey := fmt.Sprintf("session:%s", sessionID)

	// セッションデータ取得
	sessionData, err := m.client.HGetAll(ctx, sessionKey).Result()
	if err != nil {
		return err
	}

	// ユーザーセッションリストから削除
	if userID, ok := sessionData["user_id"]; ok {
		userSessionsKey := fmt.Sprintf("user_sessions:%s", userID)
		m.client.SRem(ctx, userSessionsKey, sessionID)
	}

	// セッション削除
	return m.client.Del(ctx, sessionKey).Err()
}

func (m *RedisSessionManager) GetActiveSessionsCount() (int64, error) {
	pattern := "session:*"
	keys, err := m.client.Keys(ctx, pattern).Result()
	if err != nil {
		return 0, err
	}
	return int64(len(keys)), nil
}

func main() {
	manager := NewRedisSessionManager("localhost:6379", 3600)

	// セッション作成
	userData := map[string]interface{}{
		"theme":    "dark",
		"language": "ja",
	}
	sessionID, err := manager.CreateSession(1001, userData)
	if err != nil {
		log.Fatalf("セッション作成エラー: %v", err)
	}
	fmt.Printf("Created session: %s\n", sessionID)

	// セッション取得
	sessionData, err := manager.GetSession(sessionID)
	if err != nil {
		log.Fatalf("セッション取得エラー: %v", err)
	}
	fmt.Printf("Session data: %v\n", sessionData)

	// アクティブセッション数
	count, err := manager.GetActiveSessionsCount()
	if err != nil {
		log.Fatalf("セッション数取得エラー: %v", err)
	}
	fmt.Printf("Active sessions: %d\n", count)
}

File: code/redis/examples/session_manager.py
#!/usr/bin/env python3
"""
Redis セッション管理システム

使用方法:
    from session_manager import RedisSessionManager
    
    manager = RedisSessionManager('redis-host', 6379)
    session_id = manager.create_session(1001, {'theme': 'dark'})
    session_data = manager.get_session(session_id)
"""

import redis
import json
import uuid
from datetime import datetime

class RedisSessionManager:
    def __init__(self, redis_host, redis_port, default_ttl=3600):
        self.redis_client = redis.Redis(
            host=redis_host,
            port=redis_port,
            decode_responses=True
        )
        self.default_ttl = default_ttl

    def create_session(self, user_id, user_data=None):
        """セッション作成"""
        session_id = str(uuid.uuid4())
        session_key = f"session:{session_id}"

        session_data = {
            'user_id': user_id,
            'created_at': datetime.now().isoformat(),
            'last_accessed': datetime.now().isoformat(),
            'data': json.dumps(user_data or {})
        }

        # セッションデータ保存（Hash使用）
        self.redis_client.hset(session_key, mapping=session_data)
        self.redis_client.expire(session_key, self.default_ttl)

        # ユーザーのアクティブセッション管理（Set使用）
        user_sessions_key = f"user_sessions:{user_id}"
        self.redis_client.sadd(user_sessions_key, session_id)
        self.redis_client.expire(user_sessions_key, self.default_ttl)

        return session_id

    def get_session(self, session_id):
        """セッション取得"""
        session_key = f"session:{session_id}"
        session_data = self.redis_client.hgetall(session_key)

        if not session_data:
            return None

        # 最終アクセス時刻更新
        self.redis_client.hset(session_key, 'last_accessed', datetime.now().isoformat())
        self.redis_client.expire(session_key, self.default_ttl)

        return session_data

    def update_session(self, session_id, data):
        """セッションデータ更新"""
        session_key = f"session:{session_id}"
        self.redis_client.hset(session_key, 'data', json.dumps(data))
        self.redis_client.expire(session_key, self.default_ttl)

    def destroy_session(self, session_id):
        """セッション削除"""
        session_key = f"session:{session_id}"
        session_data = self.redis_client.hgetall(session_key)

        if session_data and 'user_id' in session_data:
            user_sessions_key = f"user_sessions:{session_data['user_id']}"
            self.redis_client.srem(user_sessions_key, session_id)

        self.redis_client.delete(session_key)

    def get_active_sessions_count(self):
        """アクティブセッション数取得"""
        pattern = "session:*"
        return len(list(self.redis_client.scan_iter(match=pattern)))


# 使用例
if __name__ == "__main__":
    session_manager = RedisSessionManager('localhost', 6379)
    
    # セッション作成
    session_id = session_manager.create_session(1001, {'theme': 'dark', 'language': 'ja'})
    print(f"Created session: {session_id}")
    
    # セッション取得
    session_data = session_manager.get_session(session_id)
    print(f"Session data: {session_data}")
    
    # アクティブセッション数
    print(f"Active sessions: {session_manager.get_active_sessions_count()}")
